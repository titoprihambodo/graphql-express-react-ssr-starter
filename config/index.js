import chalk from 'chalk';

// environment checking
export const isDev = process.env.NODE_ENV === 'development' || false;
export const isProd = process.env.NODE_ENV === 'production' || false;
// webpack port for development
export const WEBPACK_PORT = !isNaN(Number(process.env.PORT)) ? Number(process.env.PORT) + 1 : 4001;

console.log(
  chalk.cyanBright(`
    Load Config
      -> environment : ${process.env.NODE_ENV} (${isProd})
      -> mongoURL : ${process.env.MONGO_URL}
      -> port : ${process.env.PORT}
      -> secret : ${process.env.SECRET}
  `)
);

export const mongoDev = 'mongodb://localhost:9999/tto-dev';
const mongoTest = 'mongodb://localhost:9999/tto-test';

// General config
const config = {
  mongoURL: isProd ? (process.env.MONGO_URL) : (isDev ? mongoDev : mongoTest),
  port: process.env.PORT || 4000,
  secret: isProd ? process.env.SECRET : process.env.SECRET || 'YfFsN998V8NSqqWAtpayJLoEvvGRkmS07MgNKx2X7jIWabiep90yJrBLK3G3GsGryQnLpxGE3dWGV62dPM61lg==',
};

export default config;