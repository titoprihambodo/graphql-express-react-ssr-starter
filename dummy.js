import Post from './models/post';
import chalk from 'chalk';

export default async () => {
  await Post.countDocuments().exec((err, count) => {
    if (count > 0) {
      return;
    }

    const text1 = 'But I must explain to you how all this mistaken idea of denouncing pleasure and praising pain was born and I will give you a complete account of the system, and expound the actual teachings of the great explorer of the truth, the master-builder of human happiness. No one rejects, dislikes, or avoids pleasure itself, because it is pleasure, but because those who do not know how to pursue pleasure rationally encounter consequences that are extremely painful. Nor again is there anyone who loves or pursues or desires to obtain pain of itself, because it is pain, but because occasionally circumstances occur in which toil and pain can procure him some great pleasure. To take a trivial example, which of us ever undertakes laborious physical exercise, except to obtain some advantage from it? But who has any right to find fault with a man who chooses to enjoy a pleasure that has no annoying consequences, or one who avoids a pain that produces no resultant pleasure? On the other hand, we denounce with righteous indignation and dislike men who are so beguiled and demoralized by the charms of pleasure of the moment, so blinded by desire';

    const text2 = 'Far far away, behind the word mountains, far from the countries Vokalia and Consonantia, there live the blind texts. Separated they live in Bookmarksgrove right at the coast of the Semantics, a large language ocean. A small river named Duden flows by their place and supplies it with the necessary regelialia. It is a paradisematic country, in which roasted parts of sentences fly into your mouth. Even the all-powerful Pointing has no control about the blind texts it is an almost unorthographic life One day however a small line of blind text by the name of Lorem Ipsum decided to leave for the far World of Grammar. The Big Oxmox advised her not to do so, because there were thousands of bad Commas, wild Question Marks and devious Semikoli, but the Little Blind Text didn’t listen. She packed her seven versalia, put her initial into the belt and made herself on the way. When she reached the first hills of the Italic Mountains, she had a last view back on the skyline of her hometown Bookmarksgrove, the headline of Alphabet Village and the subline of her own road, the Line Lane. Pityful a rethoric question ran over her cheek';

    const text3 = 'There are many variations of passages of Lorem Ipsum available, but the majority have suffered alteration in some form, by injected humour, or randomised words which don\'t look even slightly believable. If you are going to use a passage of Lorem Ipsum, you need to be sure there isn\'t anything embarrassing hidden in the middle of text. All the Lorem Ipsum generators on the Internet tend to repeat predefined chunks as necessary, making this the first true generator on the Internet. It uses a dictionary of over 200 Latin words, combined with a handful of model sentence structures, to generate Lorem Ipsum which looks reasonable. The generated Lorem Ipsum is therefore always free from repetition, injected humour, or non-characteristic words etc.';

    const post1 = new Post({
      title: 'Cicero (en)',
      content: text1,
      slug: 'cicero-dummy',
      meta: {
        description: 'Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Donec quam felis, ultricies nec, pellentesque eu, pretium quis, sem.',
        keywords: 'lorem, ipsum, dolor, sir',
      },
      createdAt: new Date("2019-01-01T12:00:00Z"),
      updatedAt: Date.now(),
    });
    const post2 = new Post({
      title: 'Far far away',
      content: text2,
      slug: 'far-far-away-dummy',
      meta: {
        description: 'Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Donec quam felis, ultricies nec, pellentesque eu, pretium quis, sem.',
        keywords: 'lorem, ipsum, dolor, sir',
      },
      createdAt: new Date("2019-01-02T12:00:00Z"),
      updatedAt: Date.now(),
    });
    const post3 = new Post({
      title: 'Where can I get some?',
      content: text3,
      slug: 'where-can-i-get-some',
      meta: {
        description: 'Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Donec quam felis, ultricies nec, pellentesque eu, pretium quis, sem.',
        keywords: 'lorem, ipsum, dolor, sir',
      },
      createdAt: new Date("2019-01-03T12:00:00Z"),
      updatedAt: Date.now(),
    });
    const post4 = new Post({
      title: 'New Dummy Post 1',
      content: text1,
      slug: 'new-post-1',
      meta: {
        description: 'Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Donec quam felis, ultricies nec, pellentesque eu, pretium quis, sem.',
        keywords: 'lorem, ipsum, dolor, sir',
      },
      createdAt: new Date("2019-01-04T12:00:00Z"),
      updatedAt: Date.now(),
    });
    const post5 = new Post({
      title: 'New Dummy Post 2',
      content: text2,
      slug: 'new-post-2',
      meta: {
        description: 'Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Donec quam felis, ultricies nec, pellentesque eu, pretium quis, sem.',
        keywords: 'lorem, ipsum, dolor, sir',
      },
      createdAt: new Date("2019-01-05T12:00:00Z"),
      updatedAt: Date.now(),
    });
    const post6 = new Post({
      title: 'New Dummy Post 3',
      content: text3,
      slug: 'new-post-3',
      meta: {
        description: 'Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Donec quam felis, ultricies nec, pellentesque eu, pretium quis, sem.',
        keywords: 'lorem, ipsum, dolor, sir',
      },
      createdAt: new Date("2019-01-06T12:00:00Z"),
      updatedAt: Date.now(),
    });
    const post7 = new Post({
      title: 'New Dummy Post 4',
      content: text1,
      slug: 'new-post-4',
      meta: {
        description: 'Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Donec quam felis, ultricies nec, pellentesque eu, pretium quis, sem.',
        keywords: 'lorem, ipsum, dolor, sir',
      },
      createdAt: new Date("2019-01-07T12:00:00Z"),
      updatedAt: Date.now(),
    });

    Post.create([post1, post2, post3, post4, post5, post6, post7], (err) => {
      if (!err) {
        console.log(
          chalk.green(`-> Post data (dummies), inserted to DB! <-`)
        )
      } else {
        console.log(
          chalk.red(`-> Post collection seems empty, but failed on inserting dummy to db. <- \n`)
        )
        console.log('​dummy -> err', err)
      }
    })
  });
}