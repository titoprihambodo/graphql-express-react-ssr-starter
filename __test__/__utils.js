import {ApolloServer} from 'apollo-server-express';
import typeDefs from '../typeDefs';
import resolvers from '../resolvers';
import models from '../models';
import jwt from 'jsonwebtoken';
import config from '../config';

export const constructTestServer = (addContext = {}) => {
  const server = new ApolloServer({
    typeDefs,
    resolvers,
    context: {
      models,
      ...addContext,
    },
  });
  return server;
};

export const whoami = async token => {
  if (token) {
    try {
      return await jwt.decode(token);
      // return await jwt.verify(token, config.secret);
    } catch (error) {
      console.log(`authentication failed, token problem or expired : ${error}`);
    }
  }
}

export const isTokenExpired = async token => {
  try {
    await jwt.verify(token, config.secret);
    return true;
  } catch (error) {
    console.log(`authentication failed, token problem or expired : ${error}`);
    return false;
  }
}