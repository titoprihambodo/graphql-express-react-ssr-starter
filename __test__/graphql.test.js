import {createTestClient} from 'apollo-server-testing';
import gql from 'graphql-tag';
import mongoose from 'mongoose';
import config from '../config';
import dummy from '../dummy';
import {constructTestServer, whoami} from './__utils';

beforeEach(async (done) => {
  if (mongoose.connection.readyState === 0) {
    await mongoose.connect(config.mongoURL, {
      useCreateIndex: true,
      useNewUrlParser: true,
    });
    await dummy();
    return done();
  }
  return done();
});

const GET_POSTS = gql`
  {
    posts {
      title
      slug
      content
      createdAt
    }
  }
`;
const GET_POST = gql`
  query ($slug: String!) {
    post(slug: $slug) {
      title
      slug
      content
      createdAt
    }
  }
`;
const ADD_POST = gql`
  mutation ($title: String!, $content: String!, $slug: String!) {
    addPost(title: $title, content: $content, slug: $slug) {
      title
      slug
      content
      createdAt
    }
  }
`;
const USER_SIGN_UP = gql`
  mutation ($username: String!, $email: String!, $password: String!) {
    signUp(username: $username, email: $email, password: $password) {
      token
    }
  }
`;
const GET_ME = gql`
  {
    me {
      username
      email
    }
  }
`;
const GET_USER_BY_USERNAME = gql`
  query ($username: String!) {
    userByUsername(username: $username) {
      username
      email
    }
  }
`;

describe('[ApolloGraphQL Server] Queries', () => {
  it('get posts', async () => {
    const server = constructTestServer();
    const {query} = createTestClient(server);
    const posts = await query({query: GET_POSTS});
    expect(posts).toBeDefined();
  });
  it('get post', async () => {
    const server = constructTestServer();
    const {query} = createTestClient(server);
    const post = await query({query: GET_POST, variables: {
      slug: 'cicero-dummy'
    }});
    expect(post).toMatchSnapshot();
  });
  it('get (me) by token', async () => {
    const token = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6IjVjOGU1OGNhYmEyMDNlMTVkM2NmYmM2NiIsImVtYWlsIjoicHJpaGFtYm9kby50aXRvQGdtYWlsLmNvbSIsInVzZXJuYW1lIjoidGl0byIsImlhdCI6MTU1MjgzMjcxNCwiZXhwIjoxNTUyODU3OTE0fQ.XWWB2bgqLing47qjKDBDipAF06V2PkY7FzL45LtaaRs';
    const me = await whoami(token);
    const server = constructTestServer({me});
    const {query} = createTestClient(server);
    const whoamiToServer = await query({query: GET_ME});
    expect(whoamiToServer).toMatchSnapshot();
  });
});

describe('[ApolloGraphQL Server] Mutation', () => {
  it('user SignUp', async () => {
    const server = constructTestServer({secret: config.secret});
    const {query, mutate} = createTestClient(server);
    const res = await mutate({mutation: USER_SIGN_UP, variables: {
      username: 'tito',
      email: 'prihambodo.tito@gmail.com',
      password: 'demotesting',
    }});
    const findRegistered = await query({query: GET_USER_BY_USERNAME, variables: {username: 'tito'}});
    expect(findRegistered).toMatchSnapshot();
  });
  it('add post', async () => {
    const server = constructTestServer();
    const {query, mutate} = createTestClient(server);
    const post = await mutate({mutation: ADD_POST, variables: {
      title: 'Dummy Post',
      content: 'Apparently we had reached a great height in the atmosphere, for the sky was a dead black, and the stars had ceased to twinkle. By the same illusion which lifts the horizon of the sea to the level of the spectator on a hillside, the sable cloud beneath was dished out, and the car seemed to float in the middle of an immense dark sphere, whose upper half was strewn with silver. Looking down into the dark gulf below, I could see a ruddy light streaming through a rift in the clouds.',
      slug: 'dummy-content',
    }});
    const added = await query({query: GET_POST, variables: {slug: 'dummy-content'}});
    expect(added.data).toMatchSnapshot();
  });
});