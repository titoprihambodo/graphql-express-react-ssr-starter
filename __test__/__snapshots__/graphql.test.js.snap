// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[ApolloGraphQL Server] Mutation add post 1`] = `
Object {
  "post": null,
}
`;

exports[`[ApolloGraphQL Server] Mutation user SignUp 1`] = `
Object {
  "data": Object {
    "userByUsername": Object {
      "email": "prihambodo.tito@gmail.com",
      "username": "tito",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`[ApolloGraphQL Server] Queries get (me) by token 1`] = `
Object {
  "data": Object {
    "me": null,
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`[ApolloGraphQL Server] Queries get post 1`] = `
Object {
  "data": Object {
    "post": null,
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
